### PR TITLE
fix(federation): remove duplicate types in schema #1001

### PR DESCRIPTION
**Pull request by dotellie**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #1001

## What is the new behavior?
This fix goes back to using `transformSchema` as used before 7.4.4 but because of the bug present in the official Apollo package where federation properties aren't carried over, this PR has that entire file copied over locally with a somewhat hacky fix. Unless Apollo suddenly changes federation however, the fix should be stable. The best solution is of course that Apollo patches their package and we can revert to using that, but I think for now this change is good to have, just so all the problems that have been popping up (such as #1001 and #981) can be fixed.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

I added some tests for these cases so they don't break again :blush:

### CODE REVIEW

**Comment by dotellie** [comment]
This PR took a while to get up because of the GitHub outage today, but here we are finally. I just noticed that I forgot to remove unused imports in the factory file (I'm used to having eslint complaining at me for that :sweat_smile:) but I don't exactly know how I should clear that up. I'm thinking of squashing both the current commits and doing that change at the same time. Thoughts?

**Comment by dotellie** [comment]
@kamilmysliwiec Any chance you could take a look at this soon?

File: lib/utils/transform-schema.util.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
+// This file is copied from `apollo-tooling`. The only difference is that it has a hack to not remove federation specific properties.
+// https://github.com/apollographql/apollo-tooling/blob/master/packages/apollo-graphql/src/schema/transformSchema.ts
```
Comment (kamilmysliwiec) [inline lib/utils/transform-schema.util.ts:2]: Can we add a note of what lines specifically must have been changed?

File: lib/federation/graphql-federation.factory.ts [hunk 2]
```diff
@@ -199,18 +201,10 @@ export class GraphQLFederationFactory {
...
-      } else if (isScalarType(type) && type.name === 'DateTime') {
-        const autoGeneratedScalar = autoGeneratedSchema.getTypeMap()[
-          key
-        ] as GraphQLScalarType;
-        if (!autoGeneratedScalar) {
-          return;
-        }
-        type.parseLiteral = autoGeneratedScalar.parseLiteral;
-        type.parseValue = autoGeneratedScalar.parseValue;
+        return type;
```
Comment (kamilmysliwiec) [inline lib/federation/graphql-federation.factory.ts:204]: What about scalars?

**Comment by kamilmysliwiec** [comment]
Thanks @dotellie! I've left 2 minor comments

**Comment by dotellie** [comment]
@kamilmysliwiec I think both your comments should be fixed now although I'm not completely sure about the scalar problem since I didn't understand it to begin with and there are no test cases for it.

**Comment by kamilmysliwiec** [comment]
Thank you @dotellie!

> I'm not completely sure about the scalar problem since I didn't understand it to begin with and there are no test cases for it

It basically inherits the `DateTime` scalar from the automatically generated schema :)

**Pull request merged**